File: internal/under/under_test.go
package under

import "testing"

func TestPathUnder(t *testing.T) {
	if !PathUnder("/repo", "/repo") || !PathUnder("/repo", "/repo/src/app.ts") {
		t.Fatal("inside")
	}
	if PathUnder("/repo", "/other") || PathUnder("/repo", "/repo-extra") || PathUnder("/repo", "/repo2") {
		t.Fatal("prefix")
	}
	if PathUnder("", "/repo") || PathUnder("/repo", "") {
		t.Fatal("empty")
	}
	if !PathUnder(`\repo`, `\repo\src`) {
		t.Fatal("slash")
	}
	if PathUnder("/", "") {
		t.Fatal("root empty")
	}
}

func TestPathsOverlap(t *testing.T) {
	if !PathsOverlap("/a", "/a/b") || !PathsOverlap("/a/b", "/a") || PathsOverlap("/a", "/b") {
		t.Fatal("overlap")
	}
}

func TestUniquePushUnder(t *testing.T) {
	list := []string{}
	if !UniquePush(&list, "/repo/a") || UniquePush(&list, "/repo/a") {
		t.Fatal("push")
	}
	UniquePush(&list, "/repo/b")
	if len(list) != 2 {
		t.Fatalf("%v", list)
	}
	got := UniqueUnder("/repo", []string{"/repo/src", "/tmp/x", "/repo/src"}, []string{"/repo/lib"})
	if len(got) != 2 || got[0] != "/repo/src" || got[1] != "/repo/lib" {
		t.Fatalf("%v", got)
	}
	if len(UniqueUnder("/repo")) != 0 {
		t.Fatal("empty")
	}
}

func TestNormalizeSlash(t *testing.T) {
	if NormalizeSlash(`C:\a\b`) != "C:/a/b" {
		t.Fatal("norm")
	}
}

File: tests/under.test.ts
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import {
  normalizeSlash,
  pathUnder,
  pathsOverlap,
  uniquePush,
  uniqueUnder,
} from "../lib/under.ts";

describe("pathUnder", () => {
  it("accepts the root and nested paths", () => {
    assert.equal(pathUnder("/repo", "/repo"), true);
    assert.equal(pathUnder("/repo", "/repo/src/app.ts"), true);
  });

  it("rejects sibling prefixes and empties", () => {
    assert.equal(pathUnder("/repo", "/repo-extra"), false);
    assert.equal(pathUnder("/repo", "/other"), false);
    assert.equal(pathUnder("", "/repo"), false);
    assert.equal(pathUnder("/repo", ""), false);
  });
});

describe("pathsOverlap / uniqueUnder", () => {
  it("overlaps nested paths", () => {
    assert.equal(pathsOverlap("/a", "/a/b"), true);
    assert.equal(pathsOverlap("/a", "/b"), false);
  });

  it("pushes unique items and filters outsiders", () => {
    const list: string[] = [];
    assert.equal(uniquePush(list, "/repo/a"), true);
    assert.equal(uniquePush(list, "/repo/a"), false);
    uniquePush(list, "/repo/b");
    assert.deepEqual(list, ["/repo/a", "/repo/b"]);
    assert.deepEqual(uniqueUnder("/repo", [["/repo/src", "/tmp/x", "/repo/src"], ["/repo/lib"]]), [
      "/repo/src",
      "/repo/lib",
    ]);
  });

  it("normalizes slashes", () => {
    assert.equal(normalizeSlash(String.raw`C:\a\b`), "C:/a/b");
  });
});
